zebra-state/src/service/check/tests/utxo.rs

```diff
@@ -0,0 +1,663 @@
+//! Randomised property tests for UTXO contextual validation
+
+use std::{convert::TryInto, sync::Arc};
+
+use proptest::prelude::*;
+
+use zebra_chain::{
+    block::{Block, Height},
+    fmt::TypeNameToDebug,
+    serialization::ZcashDeserializeInto,
+    transaction::{LockTime, Transaction},
+    transparent,
+};
+
+use crate::{
+    service::{
+        arbitrary::{new_state_with_mainnet_genesis, transaction_v4_from_coinbase},
+        StateService,
+    },
+    tests::Prepare,
+    FinalizedBlock,
+    ValidateContextError::{DuplicateTransparentSpend, MissingTransparentOutput},
+};
+
+// These tests use the `Arbitrary` trait to easily generate complex types,
+// then modify those types to cause an error (or to ensure success).
+//
+// We could use mainnet or testnet blocks in these tests,
+// but the differences shouldn't matter,
+// because we're only interested in spend validation,
+// (and passing various other state checks).
+
+proptest! {
+    /// Make sure an arbitrary transparent spend from a previous transaction in this block
+    /// is accepted by state contextual validation.
+    ///
+    /// This test makes sure there are no spurious rejections that might hide bugs in the other tests.
+    /// (And that the test infrastructure generally works.)
+    ///
+    /// It also covers a potential edge case where later transactions can spend outputs
+    /// of previous transactions in a block, but earlier transactions can not spend later outputs.
+    #[test]
+    fn accept_arbitrary_transparent_spend_from_this_block(
+        output in TypeNameToDebug::<transparent::Output>::arbitrary(),
+        mut prevout_input in TypeNameToDebug::<transparent::Input>::arbitrary_with(None),
+        use_finalized_state in any::<bool>(),
+    ) {
+        zebra_test::init();
+
+        let mut block1 = zebra_test::vectors::BLOCK_MAINNET_1_BYTES
+            .zcash_deserialize_into::<Block>()
+            .expect("block should deserialize");
+
+        // create an output
+        let output_transaction = transaction_v4_with_transparent_data([], [output.0]);
+
+        // create a spend
+        let expected_outpoint = transparent::OutPoint { hash: output_transaction.hash(), index: 0 };
+        prevout_input.set_outpoint(expected_outpoint);
+        let spend_transaction = transaction_v4_with_transparent_data([prevout_input.0], []);
+
+        // convert the coinbase transaction to a version that the non-finalized state will accept
+        block1.transactions[0] = transaction_v4_from_coinbase(&block1.transactions[0]).into();
+
+        block1
+            .transactions
+            .push(output_transaction.into());
+        block1
+            .transactions
+            .push(spend_transaction.into());
```

Comment:
Nit: this is something that I thought of, but it not really necessary: maybe it would reduce the chance of ambiguity for when different blocks are used if `extend` is used for the same block?

```suggestion
        block1
            .transactions
            .extend([output_transaction.into(), spend_transaction.into()]);
```